timemachine/fe/geometry.py

```diff
@@ -0,0 +1,67 @@
+# Utility functions to help assign and identify local geometry points
+
+from enum import Enum
+
+from rdkit import Chem
+
+
+class LocalGeometry(Enum):
+    G1_TERMINAL = 0  # R-X
+    G2_KINK = 1  # R-X-H
+    G2_LINEAR = 2  # R-X#N
+    G3_PLANAR = 3  # R-X(=O)O
+    G3_PYRAMIDAL = 4  # R-X(-H)H
+    G4_TETRAHEDRAL = 5  # R-X(-H)(-H)H
+
+
+def assign_atom_geometry(atom):
+    """
+    Heuristic using hybridization information to assign local description
+    of geometry.
+    """
+    nbrs = list(atom.GetNeighbors())
+    if len(nbrs) == 0:
+        assert 0, "Ion not supported"
+    elif len(nbrs) == 1:
+        return LocalGeometry.G1_TERMINAL
+    elif len(nbrs) == 2:
+        if atom.GetHybridization() == Chem.HybridizationType.SP3:
+            return LocalGeometry.G2_KINK
+        elif atom.GetHybridization() == Chem.HybridizationType.SP2:
+            return LocalGeometry.G2_KINK
+        elif atom.GetHybridization() == Chem.HybridizationType.SP:
+            return LocalGeometry.G2_LINEAR
+        else:
+            assert 0, "Unknown 2-nbr geometry!"
+    elif len(nbrs) == 3:
+        if atom.GetHybridization() == Chem.HybridizationType.SP3:
+            return LocalGeometry.G3_PYRAMIDAL
+        elif atom.GetHybridization() == Chem.HybridizationType.SP2:
+            return LocalGeometry.G3_PLANAR
+        else:
+            assert 0, "Unknown 3-nbr geometry"
+    elif len(nbrs) == 4:
+        if atom.GetHybridization() == Chem.HybridizationType.SP3:
+            return LocalGeometry.G4_TETRAHEDRAL
+    else:
+        assert 0, "Too many neighbors"
+
+
+def classify_geometry(mol):
+    """
+    Identify the local geometry of the molecule. This current uses a heuristic but we
+    should really be generating this from gas-phase simulations of the real forcefield.
```

Comment:
> Are there any implications for the correctness of the intended restraints if this heuristic and the forcefield disagree? (e.g. if hybridization says an atom should be pyramidal, but for some reason the forcefield makes it planar)

This is an excellent point, and one where I'm not sure how to fully resolve yet. The tests were written against molecules whose geometries are "fairly well-understood", and the current typing system is consistent with them. If we change this code to use forcefield parameters, and the tests fail, this indicates that our forcefield has issues and they should be fixed. While the restraints themselves depend on a correct description of the local geometry (and should be forcefield independent), if say right now, we misparameterized an improper torsion for sp2 atoms, then we'd have blow-ups in our simulations.

> If these should be in agreement, could you comment on how to compare these classes with the applied force field parameters? (Not immediately obvious to me, since the forcefield can express geometry preference that don't neatly fall into these classes (what if it's "partially planar" "partially pyramidal" etc.))

We'd need to run vacuum simulations and compare the local geometries computed by looking at the coordinates. The results don't have to fall "neatly", but there are certain singularities to look out for. This can be problematic where an exceeding small fraction of the time you adopt a numerically unstable geometry for the constraint (eg. H-O-H 99% of the time is in the kink formation, but it by pure luck adopted the linear formation, then any cross-product based restraints defined over H-O and O-H would blow up. And this is the most important point, so what matters is not that the angle is 90, 120, 145, etc. but rather that it's *not* 180.

Reply:
>The tests were written against molecules whose geometries are "fairly well-understood", and the current typing system is consistent with them.

Makes sense!

>If we change this code to use forcefield parameters, and the tests fail, this indicates that our forcefield has issues and they should be fixed. 

May be independently useful to check applied parameters against these classes!

(Perhaps at runtime, or perhaps on a challenging set like [vehicle](https://pubs.acs.org/doi/abs/10.1021/jm801513z).)

>The results don't have to fall "neatly", but there are certain singularities to look out for. This can be problematic where an exceeding small fraction of the time you adopt a numerically unstable geometry for the constraint (eg. H-O-H 99% of the time is in the kink formation, but it by pure luck adopted the linear formation, then any cross-product based restraints defined over H-O and O-H would blow up. And this is the most important point, so what matters is not that the angle is 90, 120, 145, etc. but rather that it's not 180.

Thanks -- this clarifies!